src/mam4xx/coagulation.hpp

```diff
@@ -59,7 +64,1120 @@ class Coagulation {
+  Real decay_const = bijqnumj1 + bijqnumj2;
+
+  constexpr float epsilonx2 = std::numeric_limits<float>::epsilon() * 2.0;
+  Real decay_factor =
+      deltat * decay_const; // calculate coag-induced changes only when this
+                            // number is not ~= zero
+
+  if (decay_factor > epsilonx2) {
+
+    // Portions of mass going into different modes
+    const Real prtn2 = bijqnumj2 / decay_const;
+    const Real prtn1 = 1.0 - prtn2;
+    const Real tmp_xf =
+        1.0 - haero::exp(-decay_factor); // total fraction lost from aitken mode
+    for (int iaer = 0; iaer < num_aer; ++iaer) {
+      const Real tmp_dq =
+          tmp_xf * qaer_bgn[iaer][nait]; // total amount lost from aitken mode
+      qaer_end[iaer][nait] -= tmp_dq;    // subtract from aitken mode
+      qaer_end[iaer][nacc] +=
+          tmp_dq * prtn1; // add a portion to accumulation mode
+      qaer_end[iaer][npca] += tmp_dq * prtn2; // add a portion to pca mode
+
+      // prtn2 (pair 3) corresponds to mass transfer to pca mode, which will
+      // lead to aging. Add this amount to the total mass gained by pca mode, to
+      // be used in the aging parameterization.
+      qaer_del_coag_out[iaer][Coagulation::i_agepair_pca] += tmp_dq * prtn2;
+    }
+  }
+
+  // --------------------------------------------------------------------
+  //  Mass transfer out of pcarbon mode. Only one coag pair is involved:
+  // - coag pair 2: pca + accumulation -> accumulation
+  // --------------------------------------------------------------------
+  decay_const = haero::max(
+      0.0, ybetaij3[1] * qnum_tavg[nacc]); // there is only 1 destination
+
+  decay_factor = deltat * decay_const; // calculate coag-induced changes only
+                                       // when this number is not ~= zero
+  if (decay_factor > epsilonx2) {
+    const Real tmp_xf =
+        1.0 - haero::exp(-decay_factor); // total fraction lost from pca mode
+    for (int iaer = 0; iaer < num_aer; ++iaer) {
+      const Real tmp_dq =
+          tmp_xf * qaer_bgn[iaer][npca]; // total amount lost from pca mode
+      qaer_end[iaer][npca] -= tmp_dq;    //  subtract from pca mode
+      qaer_end[iaer][nacc] += tmp_dq;    // add to accumulaiton mode
+    }
+  }
+}
+
+// --------------------------------------------------------------------
+// Purpose: update the number mixing ratio of a single mode considering both the
+// intramodal
+//         and intermodal coagulation
+// --------------------------------------------------------------------
+KOKKOS_INLINE_FUNCTION
+void update_qnum_for_intra_and_intermodal_coag(const Real bijdtqnumj,
+                                               const Real biidt,
+                                               const Real qnumi_bgn,
+                                               Real &qnumi_end) {
+
+  const Real eps_denominator = 1.0e-5;
+  if (bijdtqnumj < eps_denominator) {
+    qnumi_end = qnumi_bgn / (1.0 + (bijdtqnumj + biidt * qnumi_bgn) *
```

Comment:
@jeff-cohere for single precision how is  1.0 define?   double or single precision?

Reply:
Every floating point literal without a suffix in C++ is defined in double precision. https://en.cppreference.com/w/cpp/language/floating_literal